frappe/deprecation_dumpster.py

```diff
@@ -0,0 +1,290 @@
+	return text
+
+
+class Color:
+	RED = 91
+	YELLOW = 93
+	CYAN = 96
+
+
+try:
+	# since python 3.13, PEP 702
+	from warnings import deprecated as _deprecated
+except Exception:
+	import functools
+	import warnings
+	from collections.abc import Callable
+	from typing import Optional, TypeVar, Union, overload
+
+	T = TypeVar("T", bound=Callable)
+
+	def _deprecated(message: str, category=DeprecationWarning, stacklevel=1) -> Callable[[T], T]:
+		def decorator(func: T) -> T:
+			@functools.wraps(func)
+			def wrapper(*args, **kwargs):
+				if message:
+					warning_msg = f"{func.__name__} is deprecated.\n{message}"
+				else:
+					warning_msg = f"{func.__name__} is deprecated."
+				warnings.warn(warning_msg, category=category, stacklevel=stacklevel + 1)
+				return func(*args, **kwargs)
+
+			return wrapper
+			wrapper.__deprecated__ = True  # hint for the type checker
+
+		return decorator
+
+
+def deprecated(original: str, marked: str, graduation: str, msg: str, stacklevel: int = 1):
+	"""Decorator to wrap a function/method as deprecated.
+
+	Arguments:
+	        - original: frappe.utils.make_esc  (fully qualified)
+	        - marked: 2024-09-13  (the date it has been marked)
+	        - graduation: v17  (generally: current version + 2)
+	"""
+
+	def decorator(func):
+		# Get the filename of the caller
+		frame = inspect.currentframe()
+		caller_filepath = frame.f_back.f_code.co_filename
+		if os.path.basename(caller_filepath) != "deprecation_dumpster.py":
+			raise RuntimeError(
+				colorize("The deprecated function ", Color.YELLOW)
+				+ colorize(func.__name__, Color.CYAN)
+				+ colorize(" can only be called from ", Color.YELLOW)
+				+ colorize("frappe/deprecation_dumpster.py\n", Color.CYAN)
+				+ colorize("Move the entire function there and import it back via adding\n ", Color.YELLOW)
+				+ colorize(f"from frappe.deprecation_dumpster import {func.__name__}\n", Color.CYAN)
+				+ colorize("to file\n ", Color.YELLOW)
+				+ colorize(caller_filepath, Color.CYAN)
+			)
+
+		return functools.wraps(func)(
+			_deprecated(
+				colorize(f"`{original}`", Color.CYAN)
+				+ colorize(
+					f" was moved (DATE: {marked}) to frappe/deprecation_dumpster.py"
+					f" for removal (from {graduation} onwards); note:\n ",
+					Color.RED,
+				)
+				+ colorize(f"{msg}\n", Color.YELLOW),
+				stacklevel=stacklevel,
+			)
+		)(func)
+
+	return decorator
+
+
+def deprecation_warning(marked: str, graduation: str, msg: str):
+	"""Warn in-place from a deprecated code path, for objects use `@deprecated` decorator from the deprectation_dumpster"
+
+	Arguments:
+	        - marked: 2024-09-13  (the date it has been marked)
+	        - graduation: v17  (generally: current version + 2)
+	"""
+
+	warnings.warn(
+		colorize(
+			f"This codepath was marked (DATE: {marked}) deprecated"
+			f" for removal (from {graduation} onwards); note:\n ",
+			Color.RED,
+		)
+		+ colorize(f"{msg}\n", Color.YELLOW),
+		category=DeprecationWarning,
+		stacklevel=2,
+	)
```

Comment:
Utils being part of the dumpster seems wrong to me.

Reply:
Because of https://github.com/frappe/frappe/pull/27887#issuecomment-2371433065 (first answer), `def deprecated` needs to, `def deprecation_warning` not necessarily, but that also doesn't warrant, on it's own, to spread the api implementation.

While, generally, I agree, this is indeed a special case with a slightly different constraint, unless we'd want to spread the api across different files.